h2o-core/src/main/java/water/rapids/ast/prims/operators/AstBinOp.java

```diff
@@ -329,15 +343,47 @@ public void map(Chunk[] chks, NewChunk[] cress) {
           if (clf.vec().isString())
             for (int i = 0; i < clf._len; i++)
               cres.addNum(str_op(clf.atStr(lfstr, i), crt.atStr(rtstr, i)));
-          else
+          else if (categoricalOK() && clf.vec().isCategorical() && crt.vec().isCategorical()) {
+            // The vec with longer domain is iterated over due to categorical mapping
+            if (clf.vec().domain().length > crt.vec().domain().length) {
+              for (int i = 0; i < clf._len; i++) {
+                cres.addNum(op(clf.atd(i), alignedCategoricals[c][(int) crt.atd(i)]));
+              }
+            } else {
+              for (int i = 0; i < clf._len; i++) {
+                cres.addNum(op(alignedCategoricals[c][(int) clf.atd(i)], crt.atd(i)));
+              }
+            }
+          } else {
             for (int i = 0; i < clf._len; i++)
               cres.addNum(op(clf.atd(i), crt.atd(i)));
+          }
         }
       }
     }.doAll(lf.numCols(), Vec.T_NUM, new Frame(lf).add(rt)).outputFrame(lf._names, null);
     return cleanCategorical(lf, res); // Cleanup categorical misuse
   }
 
+  /**
+   * Produces a mapping array with indexes of the smaller pointing to the larger domain.
+   *
+   * @param longerDomain Domain to originally map from
+   * @param shorterDomain Domain to originally map to
+   * @return Cross-domain mapping as an array of primitive integers
+   */
+  private int[] alignCategoricals(String[] longerDomain, String[] shorterDomain) {
+    Arrays.sort(longerDomain);
+    Arrays.sort(shorterDomain);
```

Comment:
There is no need to sort the `shorterDomain`. In fact one of the domain should not be sorted because you need to preserve the correct mapping!

Reply:
Yes, the mutation is something I overlooked. Fixed ! Making a local copy of the longerDomain. Not sorting the other one as well (it worked out by accident, as it was already sorted before).